### PR TITLE
Allow to change the way the images are being downloaded

### PR DESCRIPTION
**Pull request by chrysanthos**
We have a case were the images cannot be downloaded using `fopen` but instead a client should be used. With this changes we will be able to change how the image is downloaded and instead use either curl (with additional headers) or something else.

Let me know if you want me to add an interface there or even another downloader (curl) as an option.

This is backwards compatible.

### CODE REVIEW

File: src/Downloaders/DefaultDownloader.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace Spatie\MediaLibrary\Downloaders;
+
+use Spatie\MediaLibrary\MediaCollections\Exceptions\UnreachableUrl;
+
+class DefaultDownloader
+{
+    public function getTempFile($url)
```
Comment (freekmurze) [inline src/Downloaders/DefaultDownloader.php:9]: Add a typehint and return type hint here

File: config/media-library.php [hunk 2]
```diff
@@ -91,6 +91,13 @@
      */
     'url_generator' => Spatie\MediaLibrary\Support\UrlGenerator\DefaultUrlGenerator::class,
 
+    /*
+     * When using the addMediaFromUrl method you may want to replace the default downloader.
+     * This is particularly useful when the url of the image is behind a firewall and
+     * need to add additional flags, possibly using curl.
+     */
+    'media_downloader' => Spatie\MediaLibrary\Downloaders\DefaultDownloader::class,
```
Comment (freekmurze) [inline config/media-library.php:99]: I don't think this functionality warrants at place this high in the config file. Let's move this to the bottom.

**Review by freekmurze — CHANGES_REQUESTED** [review]
Could you also add a tests and update the docs with an example of the new functionality.

**Comment by chrysanthos** [comment]
Thanks for the review @freekmurze. I made the fixes and added the docs. 

It seems like there are a lot of tests that already cover the whole method [here](https://github.com/spatie/laravel-medialibrary/blob/master/tests/Feature/FileAdder/IntegrationTest.php#L298-L319) and [here](https://github.com/spatie/laravel-medialibrary/blob/master/tests/Feature/FileAdder/IntegrationTest.php#L346-L392). Let me know if there is special case which is not handled and you want tested.

File: src/Downloaders/DefaultDownloader.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace Spatie\MediaLibrary\Downloaders;
+
+use Spatie\MediaLibrary\MediaCollections\Exceptions\UnreachableUrl;
+
+class DefaultDownloader
```
Comment (freekmurze) [inline src/Downloaders/DefaultDownloader.php:7]: Let's make this class implement an `Downloader` interface with that `getTempFile` method in it. In the readme and config file it should be mentioned. That a valid downloader is any class that implements that new interface.

**Review by freekmurze — CHANGES_REQUESTED** [review]

**Comment by chrysanthos** [comment]
Hey @freekmurze, does this seem ready to be merged or do you have any more changes you'd like to see?

**Comment by freekmurze** [comment]
I need to find some time to review this.

**Comment by freekmurze** [comment]
Thanks!

**Pull request merged**